nntrainer/layers/time_dist.cpp

```diff
@@ -23,6 +23,68 @@ namespace nntrainer {
 
 const std::string TimeDistLayer::type = "time_dist";
 
+static void reshape(Tensor &m) {
+  TensorDim d = m.getDim();
+  m.reshape({d[2], d[1], d[0], d[3]});
+}
+
+void TimeDistLayer::setPosition() {
+  positions[0] = net_input[0]->getVariableRef().getData();
+  positions[1] = net_input[0]->getGradientRef().getData();
+  positions[2] = net_hidden[0]->getVariableRef().getData();
+  positions[3] = net_hidden[0]->getGradientRef().getData();
+}
+
+void TimeDistLayer::transposeInOut() {
+  // Position[0] : net_input.variable
+  Tensor &input_ = net_input[0]->getVariableRef();
+  input_.copy(transposeTensor(input_));
+
+  // Position[1] : net_input.gradient
+  Tensor &ret_ = net_input[0]->getGradientRef();
+  if (ret_.getData() != positions[0]) {
+    ret_.copy(transposeTensor(ret_));
+  } else {
+    reshape(ret_);
+  }
+
+  // Position[2] : net_hidden.variable
+  Tensor &hval_ = net_hidden[0]->getVariableRef();
+  if (hval_.getData() != positions[0] && hval_.getData() != positions[1]) {
+    hval_.copy(transposeTensor(hval_));
+  } else {
+    reshape(hval_);
+  }
+
+  // Position[3] : net_hidden.gradient
+  bool trans = true;
+
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  for (unsigned int i = 0; i < 3; ++i) {
+    if (derivative_.getData() == positions[i]) {
+      trans = false;
+      break;
+    }
+  }
+  if (trans)
+    derivative_.copy(transposeTensor(derivative_));
+  else
+    reshape(derivative_);
+}
+
+Tensor TimeDistLayer::transposeTensor(Tensor &m) {
+  TensorDim dim = m.getDim();
+  // Assume the channel is 1. It transpose [b, 1, h, w] to [h, 1, b, w ]
+  // and nntrainer only support 1,2,3 transpose. So we do reshape first to make
+  // [1, b,h, w]
+  m.reshape({dim[1], dim[0], dim[2], dim[3]});
```

Comment:
how about we reshape to `{1, dim[0] * dim[1], dim[2], dim[3]}`. This will transpose to `{1, dim[2], dim[0] * dim[1], dim[3]}`, and then can be reshaped back to `{dim[2], dim[0], dim[1], dim[3]}` always, or `{dim[2], dim[1], dim[0], dim[3]}` here when dim[1] is 1.

Reply:
I will leave the comment and leave it as future work. thanks.

Reply:
cool.